File: 3575.maximum-good-subtree-score.py
#
# @lc app=leetcode id=3575 lang=python3
#
# [3575] Maximum Good Subtree Score
#

# @lc code=start
class Solution:
    def goodSubtreeSum(self, vals: List[int], par: List[int]) -> int:
        from collections import defaultdict
        MOD = 10**9 + 7
        # Build tree from parent array
        tree = defaultdict(list)
        n = len(vals)
        for i in range(1, n):
            tree[par[i]].append(i)
        
        # Helper function to get digits as a set from a number
        def get_digits(num):
            return set(str(num))
        
        # DFS function to calculate maximum good subtree score
        def dfs(node):
            current_digits = get_digits(vals[node])
            current_score = vals[node]
            for child in tree[node]:
                child_score, child_digits = dfs(child)
                if not current_digits & child_digits: # If no common digits
                    current_score += child_score
                    current_digits |= child_digits # Union of digits sets if valid
            return current_score, current_digits
        
        # Calculate maxScore for each node starting from root (0) and sum it up
        total_sum = 0
        for i in range(n): # Calculate maxScore for each subtree rooted at i
            score, _ = dfs(i)
            total_sum += score % MOD # Sum up scores with modulo operation
total_sum %= MOD # Final mod operation after summing up all scores
return total_sum

# @lc code=end

File: 3575.maximum-good-subtree-score.java
#
# @lc app=leetcode id=3575 lang=java
#
# [3575] Maximum Good Subtree Score
#

# @lc code=start
import java.util.*;

class Solution {
    static final int MOD = 1000000007;
    public int goodSubtreeSum(int[] vals, int[] par) {
        int n = vals.length;
        List<Integer>[] tree = new ArrayList[n];
        for (int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for (int i = 1; i < n; ++i) {
            tree[par[i]].add(i);
        }
        int[] maxScore = new int[n];
        dfs(0, vals, tree, maxScore);
        long sum = 0;
        for (int score : maxScore) sum = (sum + score) % MOD;
        return (int)sum;
    }

    // Returns: Map<bitmask, sum> for subtree rooted at u
    private Map<Integer, Integer> dfs(int u, int[] vals, List<Integer>[] tree, int[] maxScore) {
        Map<Integer, Integer> dp = new HashMap<>();
        int mask = digitMask(vals[u]);
        dp.put(mask, vals[u]);
        for (int v : tree[u]) {
            Map<Integer, Integer> childDP = dfs(v, vals, tree, maxScore);
            Map<Integer, Integer> newDP = new HashMap<>(dp);
            for (Map.Entry<Integer, Integer> e1 : dp.entrySet()) {
                int m1 = e1.getKey(), s1 = e1.getValue();
                for (Map.Entry<Integer, Integer> e2 : childDP.entrySet()) {
                    int m2 = e2.getKey(), s2 = e2.getValue();
                    if ((m1 & m2) == 0) {
                        int nm = m1 | m2;
                        int ns = s1 + s2;
                        newDP.put(nm, Math.max(newDP.getOrDefault(nm, 0), ns));
                    }
                }
            }
            // Add all child-only masks (for cases where we don't select current node)
            for (Map.Entry<Integer, Integer> e : childDP.entrySet()) {
                newDP.putIfAbsent(e.getKey(), e.getValue());
            }
            dp = newDP;
        }
        int max = 0;
        for (int v : dp.values()) max = Math.max(max, v);
        maxScore[u] = max;
        return dp;
    }

    // Returns bitmask of digits present in x
    private int digitMask(int x) {
        int mask = 0;
        while (x > 0) {
            mask |= 1 << (x % 10);
            x /= 10;
        }
        return mask;
    }
}
# @lc code=end